OpenRA.Mods.Common/Activities/Air/Land.cs

```diff
@@ -42,22 +46,28 @@ public override Activity Tick(Actor self)
 					return this;
 			}
 
-			if (!target.IsValidFor(self))
+			// TargetType.Invalid indicates target was an actor and either killed, removed or teleported
+			if (target.Type == TargetType.Invalid)
+			{
+				Cancel(self);
 				return NextActivity;
+			}
 
 			if (IsCanceling)
 			{
 				aircraft.RemoveInfluence();
 				return NextActivity;
 			}
 
-			if (requireSpace && !landingInitiated)
+			if (!landingInitiated)
 			{
-				var landingCell = self.World.Map.CellContaining(target.CenterPosition);
-				if (!aircraft.CanLand(landingCell, ignoreActor))
+				var landingCell = !aircraft.Info.VTOL ? self.World.Map.CellContaining(target.CenterPosition + offset) : self.Location;
+				if (!aircraft.CanLand(landingCell, target.Actor))
 				{
 					// Maintain holding pattern.
-					QueueChild(self, new FlyCircle(self, 25), true);
+					if (!aircraft.Info.CanHover)
+						QueueChild(self, new FlyCircle(self, 25), true);
```

Comment:
Isn't the actor going to be in completely the wrong position to land after this?

Reply:
Probably. But in my opinion, properly fixing issues like this requires splitting (and improving) at least the 'calculate approach vector and queue move activities' part of `ReturnToBase` to a separate activity that can be used in places like this. I can try to do that before next release, but not as part of this PR.

Reply:
Definitely work for a future PR.

The approach vector code should move here into `Land` rather than a separate activity, since "landing" is itself a logical activity that should be treated as a single unit - we ask the actor to "land in cell X" or "dock with actor Y" and expect it to sort out the details and deal with any unexpected delays or blockages without us having to micromanage it.

Harvesters are a good parallel to this, we used to have `FindAndDeliverResources` (and `FindResources` before it) manage everything, moving the actor and then queuing `Harvest` to harvest whatever is in the same cell as the actor.  @tovl's recent PRs changed this to make `Harvest` much more logical and robust: it is given the target cell, and works out the details of moving to the cell if needed before harvesting.